node/src/test_utils/recorder_stop_conditions.rs

```diff
@@ -336,10 +369,96 @@ mod tests {
+            StopConditions::Any(_) => {
+                panic!("Stage 2: expected StopConditions::AllGreedily, not Any")
+            }
+            StopConditions::AllLazily(_) => {
+                panic!("Stage 2: expected StopConditions::AllGreedily, not AllLazily")
+            }
+        }
+    }
+
+    #[test]
+    fn match_all_with_conditions_gradually_eliminated_lazily_until_empty() {
+        let mut cond_set = StopConditions::AllLazily(vec![
+            MsgIdentification::ByPredicate {
+                predicate: Box::new(|msg| {
+                    if let Some(ip_msg) = msg.downcast_ref::<NewPublicIp>() {
+                        ip_msg.new_ip.is_ipv6()
+                    } else {
+                        false
+                    }
+                }),
+            },
+            MsgIdentification::ByType(TypeId::of::<NewPublicIp>()),
+            MsgIdentification::ByType(TypeId::of::<NewPublicIp>()),
+        ]);
+        let tested_msg_1 = ScanForNewPayables {
+            response_skeleton_opt: None,
+        };
+
+        let kill_system = cond_set.resolve_stop_conditions::<ScanForNewPayables>(&tested_msg_1);
+
+        assert_eq!(kill_system, false);
+        assert_on_state_after_lazily_matched(1, &cond_set, |conds| {
+            assert_eq!(conds.len(), 3);
+            assert!(matches!(conds[0], MsgIdentification::ByPredicate { .. }));
+            assert!(matches!(conds[1], MsgIdentification::ByType(_)));
+            assert!(matches!(conds[2], MsgIdentification::ByType(_)));
+        });
+        let tested_msg_2 = NewPublicIp {
+            new_ip: IpAddr::V4(Ipv4Addr::new(6, 7, 8, 9)),
+        };
+
+        let kill_system = cond_set.resolve_stop_conditions::<NewPublicIp>(&tested_msg_2);
+
+        assert_eq!(kill_system, false);
+        assert_on_state_after_lazily_matched(2, &cond_set, |conds| {
+            assert_eq!(conds.len(), 2);
+            assert!(matches!(conds[0], MsgIdentification::ByPredicate { .. }));
+            assert!(matches!(conds[1], MsgIdentification::ByType(_)));
+        });
+        let tested_msg_3 = NewPublicIp {
+            new_ip: IpAddr::V6(Ipv6Addr::new(1, 2, 4, 1, 4, 3, 2, 1)),
+        };
+
+        let kill_system = cond_set.resolve_stop_conditions::<NewPublicIp>(&tested_msg_3);
+
+        assert_eq!(kill_system, false);
+        assert_on_state_after_lazily_matched(3, &cond_set, |conds| {
+            assert_eq!(conds.len(), 1);
+            assert!(matches!(conds[0], MsgIdentification::ByType(_)))
+        });
+        let tested_msg_4 = NewPublicIp {
+            new_ip: IpAddr::V4(Ipv4Addr::new(45, 45, 45, 45)),
+        };
+
+        let kill_system = cond_set.resolve_stop_conditions::<NewPublicIp>(&tested_msg_4);
+
+        assert_eq!(kill_system, true);
+        assert_on_state_after_lazily_matched(4, &cond_set, |conds| {
+            assert!(conds.is_empty());
+        });
```

Comment:
In case you could introduce line breaks right after a particular set of assertion ends, that would really help in dividing the stages.